akka-remote/src/main/scala/akka/remote/artery/ArteryTcpTransport.scala

```diff
@@ -0,0 +1,352 @@
+            remoteAddress,
+            halfClose = true, // issue https://github.com/akka/akka/issues/24392 if set to false
+            connectTimeout = connectionTimeout)
+      }
+
+    def connectionFlowWithRestart: Flow[ByteString, ByteString, NotUsed] = {
+      val flowFactory = () ⇒ connectionFlow.mapMaterializedValue(_ ⇒ NotUsed)
+        .log(name = s"outbound connection to [${outboundContext.remoteAddress}], ${streamName(streamId)} stream")
+        .addAttributes(Attributes.logLevels(onElement = LogLevels.Off, onFailure = Logging.WarningLevel))
+
+      // FIXME config of backoff
+      import scala.concurrent.duration._
+      if (streamId == ControlStreamId) {
+        // restart of inner connection part important in control flow, since system messages
+        // are buffered and resent from the outer SystemMessageDelivery stage.
+        // FIXME The mat value Future is currently never completed, because RestartFlow will retry forever, should it give up?
+        //       related to give-up-system-message-after ?
+        RestartFlow.withBackoff[ByteString, ByteString](1.second, 5.seconds, 0.1)(flowFactory)
+      } else {
+        // Best effort retry a few times
+        // FIXME only restart on failures?, but missing in RestartFlow, see https://github.com/akka/akka/pull/23911
+        RestartFlow.withBackoff[ByteString, ByteString](1.second, 5.seconds, 0.1, maxRestarts = 3)(flowFactory)
+      }
+    }
+
+    Flow[EnvelopeBuffer]
+      .map { env ⇒
+        val bytes = ByteString(env.byteBuffer)
+        bufferPool.release(env)
+        bytes
+      }
+      .recoverWithRetries(1, { case ArteryTransport.ShutdownSignal ⇒ Source.empty })
+      .via(connectionFlowWithRestart)
+      .map(_ ⇒ throw new IllegalStateException(s"Unexpected incoming bytes in outbound connection to [${outboundContext.remoteAddress}]"))
+      .toMat(Sink.ignore)(Keep.right)
+  }
+
+  override protected def runInboundStreams(): Unit = {
+    implicit val mat = materializer
+    implicit val sys = system
+
+    val (controlStream, controlStreamCompleted) = runInboundControlStream()
+    val (ordinaryMessagesStream, ordinaryMessagesStreamCompleted) = runInboundOrdinaryMessagesStream()
+    val (largeMessagesStream, largeMessagesStreamCompleted) = {
+      if (largeMessageChannelEnabled)
+        runInboundLargeMessagesStream()
+      else
+        (
+          Flow[EnvelopeBuffer]
+          .map(_ ⇒ log.warning("Dropping large message, missing large-message-destinations configuration."))
+          .to(Sink.ignore),
+          Promise[Done]().future) // never completed, not enabled
+    }
+
+    val inboundStream: Sink[EnvelopeBuffer, NotUsed] =
+      Sink.fromGraph(GraphDSL.create() { implicit b ⇒
+        import GraphDSL.Implicits._
+        val partition = b.add(Partition[EnvelopeBuffer](3, env ⇒ {
+          val streamId: Int = env.byteBuffer.get(EnvelopeBuffer.StreamIdOffset).toInt
+          if (streamId == OrdinaryStreamId) 1
```

Comment:
I thought this looked nicer as pattern match but only cosmetic

Reply:
I agree, I was debugging something, but it was not solved by this. Changing back, and adding a `case _ => ` for the exception